install.sh

```diff
@@ -0,0 +1,408 @@
+Installs the chs-dev CLI tool.
+
+OPTIONS
+-------
+
+-d <directory> - local directory to install CLI to (Defaults to ~/.chs-dev)
+-f - forces the installation - will clobber existing installation. Without the flag,
+    user will have to confirm reinstallation
+-l <loglevel> - sets the log level to set logging output, valid options are DEBUG,
+    INFO, WARN, ERROR (defaults to INFO)
+-S - when set does not create symlink to binary
+-s <path-to-symlink-loc> - Sets the path to the place to output the symlink to (Defaults to ~/.companies_house_config/bin)
+-v <version> - installs a specific version (Defaults to latest)
+-h print this message and exit
+
+ARGUMENTS
+---------
+
+COMMAND - Specifies the action to do, must be install or uninstall. Defaults to installl
+EOF
+}
+
+# Logs error and then exits process with error exit code
+panic() {
+  log ERROR "${1:?message required}" >&2
+  exit 1
+}
+
+# Logs messages to terminal and controls logging
+log() {
+  log_level="${1?:level required}"
+  log_message="${2:?message required}"
+
+  case "${LOGGING_LEVEL}" in
+  DEBUG) ;;
+  INFO)
+    if [ "${log_level}" = 'DEBUG' ]; then
+      return 0
+    fi
+    ;;
+  WARN*)
+    if [ "${log_level}" = 'DEBUG' ] || [ "${log_level}" = 'INFO' ]; then
+      return 0
+    fi
+    ;;
+  ERROR)
+    if [ ! "${log_level}" = "ERROR" ]; then
+      return 0
+    fi
+    ;;
+  esac
+
+  case "${log_level}" in
+  DEBUG) log_colour="" ;;
+  INFO) log_colour="${F_FG_GREEN}" ;;
+  WARN*) log_colour="${F_FG_YELLOW}" ;;
+  ERROR) log_colour="${F_FG_RED}" ;;
+  *) panic "Invalid log level: ${log_level}" ;;
+  esac
+
+  printf '[%s] - [%s] - %s\n' "${F_FG_CYAN}$(date +'%Y-%m-%dT%H:%M:%S')${F_RESET}" "${log_colour}${F_BOLD}${log_level}${F_RESET}" "${log_message}"
+}
+
+# Checks that the requested version is valid i.e. there is a release with the
+# name of the version supplied
+# RETURNS
+# 0  - version is valid
+# 1  - version is not valid
+validate_version() {
+  log DEBUG 'validating version'
+
+  {
+    jq -r '.[] | .name' "${releases_file}"
+    printf -- '%s' "$?" >"${temp_d}"/jq_status
+  } |
+    {
+      grep -q "${VERSION}"
+      printf -- '%s' "$?" >"${temp_d}"/grep_status
+    }
+
+  if ! grep -q '^0$' "${temp_d}"/jq_status ||
+    ! grep -q '^0$' "${temp_d}"/grep_status; then
+    log ERROR "Version ${VERSION} not found"
+    return 1
+  fi
+}
+
+# Sets the variable OS to the name of the operating system. Panics if the
+# operating system is unsupported by the CLI (i.e. missing compiled CLI
+# artifacts). Panics if not OSX or Linux compatible
+determine_operating_system() {
+  log DEBUG 'determining operating system'
+  uname_out="$(uname -a)"
+  case "${uname_out}" in
+  *Microsoft*) OS="linux" ;; #must be first since Windows subsystem for linux will have Linux in the name too
+  *microsoft*) OS="linux" ;; #WARNING: My v2 uses ubuntu 20.4 at the moment slightly different name may not always work
+  Linux*) OS="linux" ;;
+  Darwin*) OS="darwin" ;; # Apple Mac
```

Comment:
Making it case insensitive (i.e. lowercasing) the 2 Microsoft will be 1:
```
uname_out="$(uname -a | tr '[:upper:]' '[:lower:]')"
```

Reply:
oddly this is interfering with the output

> `Dagwin ABM-WVG6010RD0 23.4.0 Dagwin Kfgnfl Vfgsion 23.4.0b Fgi Mag 15 00b10b42 PDT 2024; gootbxnc-10063.101.17~1/RELEASE_ARM64_T6000 agm64`